### PR TITLE
Improve docs for `vcat`, `cat`, etc.

### PR DESCRIPTION
**Pull request by mcabbott**
This wants to:
* Mention the fast methods for `reduce(vcat, A)` under `reduce` and under `vcat`, and state *exactly* what types they accept. As far as I know this was documented nowhere besides `methods(reduce)`.
* Mention `stack` from [#43334](https://github.com/JuliaLang/julia/pull/43334) as an alternative where appropriate, including under `eachslice` and `mapslices`.
* Provide better examples for `vcat`, `hcat`. They are about the same length, but try not to print out things like `b = [6 7; 8 9; 10 11; 12 13; 14 15]` too often, instead illustrating more uses, from simple to complex.
* Break up the wall of text for `cat(A...; dims)`, and avoid saying the word "stacked" there.

### CODE REVIEW

**Comment by mikmoore** [comment]
Before we canonize `reduce(vcat,itr)` and friends as the proper way to do these things, should we consider whether this is actually something we want to do? I've always despised these specializations. They're difficult to discover (this doc PR helps) and they don't compose or behave like you'd hope, e.g. `reduce(vcat ∘ vec,itr)` or `mapreduce(vec,vcat,itr)` will totally miss the specializations and be slow as anything.

#21672 applies in a narrow subset of cases. #46003 would apply in some others. But really, we still lack a generic function for concatenation of iterables. `cat` provides the analog of `min` but we lack an analog of `minimum`. I closed #45563 awhile back but I don't think `stack` from #21672 really met my hopes on this matter.

**Comment by mcabbott** [comment]
These specialisations probably cannot be removed, so at least they should be clearly documented, including the precise signature which triggers them. If I'm not mistaken you have to look at  `methods(reduce)` to know this, at present.

I agree they are hard to discover, and fragile. `stack` covers only some uses. Maybe other cases are off-topic for this PR so I fold them here:

<details>

Similar code could certainly be written for an efficient eager `flatten`; in some sense that would be the `reduce(vcat, v_v)` to match `stack`'s `reduce(hcat, v_v)`, and has the precedent of `Iterators.flatten` to describe exactly what it does. Whether it's likely to be approved I don't know. 

Many other rules are possible. If I understand right, https://github.com/JuliaLang/julia/pull/46003 proposes a rule with a `kron`-like combining of the axes of inner and outer arrays:
```julia
julia> A = [rand(2,3) for _ in 1:4, _ in 1:5];

julia> awfulstack(A) == hvcat(5, permutedims(A)...)
true

julia> using TensorCast

julia> awfulstack(A) == @cast out[(i,I), (j,J)] := A[I,J][i,j]
true

julia> stack(A) == @cast out[i,j,I,J] := A[I,J][i,j]
true

julia> awfulstack(A) == @cast out[(i,I), (j,J)] := stack(A)[i,j,I,J]
true
```
On a uniform-size array of arrays like `A`, this is (like `flatten`) some `reshape` & `permutedims` of `stack`. There's also some extension to non-uniform arrays which I don't quite see. 

</details>

File: base/reducedim.jl [hunk 1]
```diff
@@ -382,6 +382,9 @@ The associativity of the reduction is implementation-dependent; if you need a pa
 associativity, e.g. left-to-right, you should write your own loop or consider using
 [`foldl`](@ref) or [`foldr`](@ref). See documentation for [`reduce`](@ref).
 
+There are special efficient methods for `reduce(`[`vcat`](@ref)`, xs)` and
+`reduce(`[`hcat`](@ref)`, xs)` for certain arrays of arrays; see also [`stack`](@ref).
```
Comment (mcabbott) [inline base/reducedim.jl:386]: This is in the wrong place, didn't mean to find `reduce(f, A)`.

File: base/abstractarray.jl [hunk 2]
```diff
@@ -3075,6 +3093,11 @@ julia> mapslices(f, A, dims=(1,2))
 [:, :, 3] =
  21  21  21  21
 
+julia> f(x::AbstractMatrix) = fill(x[1,1], 1,4);
+
+julia> B == stack(f, eachslice(A, dims=3))
+true
```
Comment (mcabbott) [inline base/abstractarray.jl:3099]: Maybe this function should have a distinct name:
```suggestion
julia> f2(x::AbstractMatrix) = fill(x[1,1], 1,4);

julia> B == stack(f2, eachslice(A, dims=3))
true
```

File: base/abstractarray.jl [hunk 3]
```diff
@@ -1794,111 +1794,118 @@ end
...
 
-julia> hcat(x, [1; 2; 3])
-3×1 Matrix{Any}:
- 1
- 2
- 3
+julia> Float32[1 2 [30 40] [5, 6]']  # syntax for supplying the eltype
+1×6 Matrix{Float32}:
+ 1.0  2.0  30.0  40.0  5.0  6.0
```
Comment (mcabbott) [inline base/abstractarray.jl:1887]: Should this mention `typed_hcat`? This is not mentioned in https://docs.julialang.org/en/v1/manual/arrays/#Typed-array-literals, only in devdocs  https://docs.julialang.org/en/v1/devdocs/ast/#Bracketed-forms  so probably not.

File: base/abstractarray.jl [hunk 4]
```diff
@@ -1794,111 +1794,118 @@ end
...
 ```
 """
 vcat(X...) = cat(X...; dims=Val(1))
 """
     hcat(A...)
 
-Concatenate along dimension 2. To efficiently concatenate a large vector of arrays,
-use `reduce(hcat, x)`.
+Concatenate arrays or numbers horizontally. Equivalent to [`cat`](@ref)`(A...; dims=2)`,
+which is also called by the syntax `[a b c]` or `[a;;]`.
```
Comment (mcabbott) [inline base/abstractarray.jl:1863]: The doc section on array literals is good: https://docs.julialang.org/en/v1/manual/arrays/#man-array-concatenation and it mentions functions `hcat` etc. Should these docstrings refer back to it?

**Comment by mikmoore** [comment]
Technically, removing these performance shortcuts would be non-breaking since the fallback is still operable. That said, reversion to the fallback would impact performance very negatively and for that reason I imagine we'd keep them for a long time anyway.

I expect that, if we do ever manage a more generic function, these specialization could still exist (for legacy reasons) although be replaced by forwarding to that function. So, with that in mind and with the lack of an in-progress PR to create the aforementioned generalization, I'll retract my reservations about documenting them.

**Comment by mcabbott** [comment]
Besides performance there are quirks like this which would probably break things:
```
julia> mapreduce(identity, hcat, [[1,2]])  # never calls hcat
2-element Vector{Int64}:
 1
 2

julia> reduce(hcat, [[1,2]])
2×1 Matrix{Int64}:
 1
 2
```
Perhaps we could make `reduce(hcat, ::AbstractVector{<:AbstractVector})` call `stack` now, but there's not much gain, and maybe there are other weird edge cases.

File: base/abstractarray.jl [hunk 5]
```diff
@@ -1801,111 +1801,118 @@ end
...
+julia> hcat(Int[], Int[], Int[])  # empty vectors, each of size (0,)
+0×3 Matrix{Int64}
+
+julia> col0 = Matrix(undef, 2, 0)  # empty matrix, size (2,0)
+2×0 Matrix{Any}
+
+julia> hcat([1.1, 9.9], col0, ms[3])
+2×3 Matrix{Any}:
+ 1.1  50  60
+ 9.9  70  80
```
Comment (mcabbott) [inline base/abstractarray.jl:1915]: This is the last test failure here:
```
│ hcat([1.1, 9.9], col0, ms[3])
--
  | │
  | │ Evaluated output:
  | │
  | │ 2×3 Matrix{Any}:
  | │  1.1  50.0  60.0
  | │  9.9  70.0  80.0
  | │
  | │ Expected output:
  | │
  | │ 2×3 Matrix{Any}:
  | │  1.1  50  60
  | │  9.9  70  80
  | │

```

**Comment by mcabbott** [comment]
Bump? Test failures look unrelated to me.

File: base/abstractarray.jl [hunk 6]
```diff
@@ -1916,19 +1923,23 @@ typed_hcat(::Type{T}, X...) where T = _cat_t(Val(2), T, X...)
...
+which will also be the size of the output array along those dimensions.
 
-See also [`hcat`](@ref), [`vcat`](@ref), [`hvcat`](@ref), [`repeat`](@ref).
+If `dims` is a single number, the different arrays are tightly packed along that dimension.
+If `dims` is an iterable containing several dimensions, the positions along these dimensions
+are increased simultaneously for each input array, filling with zero elsewhere.
+This allows one to construct block-diagonal matrices as `cat(matrices...; dims=(1,2))`,
+and their higher-dimensional analogues.
+
+The keyword also accepts `Val(dims)`.
```
Comment (mcabbott) [inline base/abstractarray.jl:1939]: This fails on 1.6 but works on 1.9 --- so perhaps there ought to be a compat note about Val somewhere? 
```
julia> VERSION
v"1.6.0"

julia> cat(1,2,dims=Val((1,2)))
ERROR: MethodError: no method matching isless(::Tuple{Int64, Int64}, ::Int64)
```

Comment (mcabbott) [reply base/abstractarray.jl:1939]:
```suggestion
The keyword also accepts `Val(dims)`.

!!! compat "Julia 1.8"
    For multiple dimensions `dims = Val(::Tuple)` was added in Julia 1.8. `dims = Val(::Int)` always worked.
```

Comment (mcabbott) [reply base/abstractarray.jl]: Done.

**Comment by mcabbott** [comment]
Bump... pinging @ViralBShah as he liked it in August?

**Comment by ViralBShah** [comment]
Thanks. I will come around to this soon (on phone now).

In general I feel anything that is a strict improvement and is correct on docs side should be merged.

**Pull request merged**